router/delivery/http/router_handler.go

```diff
@@ -85,19 +174,27 @@ func (a *RouterHandler) GetOptimalQuote(c echo.Context) (err error) {
 		// Note: we do not end the span here as it is ended in the middleware.
 	}()
 
-	isSingleRoute, err := domain.ParseBooleanQueryParam(c, "singleRoute")
-	if err != nil {
-		return c.JSON(domain.GetStatusCode(err), domain.ResponseError{Message: err.Error()})
+	var req GetQuoteRequest
+	if err := UnmarshalRequest(c, &req); err != nil {
+		return err
 	}
 
-	shouldApplyExponents, err := domain.ParseBooleanQueryParam(c, "applyExponents")
-	if err != nil {
-		return c.JSON(domain.GetStatusCode(err), domain.ResponseError{Message: err.Error()})
+	// Validate the request
+	if err := req.Validate(c); err != nil {
+		return err
 	}
 
-	tokenOutDenom, tokenIn, err := getValidRoutingParameters(c)
-	if err != nil {
-		return err
+	var (
```

Comment:
I think it is more business logic than HTTP handlers. Should we pass `GetQuoteRequest` to `tokensusecase.GetOptimalQuote` and move this logic there?

Reply:
Good point. 

I think we should keep `GetOptimalQuote` as is but rename it to `GetOptimalQuoteOutGivenIn`.

Then, add a new method`GetOptimalQuoteInGivenOut` to encapsulate the new business logic.

However, I think that the logic so far for deciding on the inputs makes sense to keep here. What I would add to `GetOptimalQuoteInGivenOut` is the logic for inverting the routes to display the final result.

We will likely also need a new implementation/handling logic for `PrepareResult`:
https://github.com/osmosis-labs/sqs/blob/16f75b936a18c3bf8ce2a1f85d3f542712c40b2c/router/usecase/quote.go#L39